contracts/adapters/IPRebalanceExtension.sol

```diff
@@ -0,0 +1,555 @@
+            "untransform unavailable"
+        );
+
+        uint256 targetUnderlying = targetUnitsUnderlying[_transformComponent];
+        uint256 currentUnits = setToken.getDefaultPositionRealUnit(_transformComponent).toUint256();
+
+        // convert target units from underlying to transformed amounts
+        uint256 exchangeRate = transformInfo.transformHelper.getExchangeRate(transformInfo.underlyingComponent, _transformComponent);
+        uint256 targetUnitsInTransformed = targetUnderlying.preciseMul(exchangeRate);
+        uint256 unitsToUntransform = currentUnits > targetUnitsInTransformed ? currentUnits.sub(targetUnitsInTransformed) : 0;
+        
+        uint256 maxUntransformUnits = transformInfo.maxTransformSize.preciseDiv(setToken.totalSupply()).preciseMul(exchangeRate);
+        if (unitsToUntransform > maxUntransformUnits) {
+            unitsToUntransform = maxUntransformUnits;
+        }
+
+
+        require(unitsToUntransform > 0, "nothing to untransform");
+
+        (address module, bytes memory callData) = transformInfo.transformHelper.getUntransformCall(
+            manager.setToken(),
+            transformInfo.underlyingComponent,
+            _transformComponent,
+            unitsToUntransform,
+            _untransformData
+        );
+
+        invokeManager(module, callData);
+    }
+
+    /**
+     * Transforms a component. To prevent losses from untransforms that may incur slippage, a maximum transform
+     * size and minimum transform delay are enforced. If _transformRemaining is set to true, ignore the calculations
+     * described below and simply transform all remaining underlying tokens.
+     *
+     * To calculate the number of units to transform use:
+     * max(targetUnitsUnderlying - currentUnits / exchangeRate, 0)
+     *
+     * This calculation ensures that if the current units (measured in underlying amounts) is greater than the target
+     * underlying units then no tokens will be transformed. If the calculated units to transform are greater than the
+     * units of underlying left in the set, just untransform all the units left. This may happen in cases where the exchange
+     * rate changes between the time the rebalance was parameterized and the transformation was made. Finally, if the amount
+     * to transform is greater than maximum transform amount, the transform size will be replaced with the max value. The rest
+     * can be transformed in later transactions.
+     * 
+     */
+    function _transform(address _transformComponent, bytes memory _transformData, bool _transformRemaining) internal {
+        require(stage == RebalanceStage.TRANSFORM, "must be in transform stage");
+
+        TransformInfo memory transformInfo = transformComponentInfo[_transformComponent];
+
+        require(transformInfo.minTransformDelay + lastTransform[_transformComponent] < block.timestamp, "delay not elapsed");
```

Comment:
All the requires can be grouped together and moved to an internal function `validateTransform`